src/libsystemd/sd-bus/bus-socket.c

```diff
@@ -156,93 +156,109 @@ static int bus_socket_write_auth(sd_bus *b) {
 }
 
 static int bus_socket_auth_verify_client(sd_bus *b) {
-        char *d, *e, *f, *start;
+        char *l, *lines[4] = {};
         sd_id128_t peer;
+        size_t i, n;
         int r;
 
         assert(b);
 
         /*
-         * We expect three response lines:
-         *   "DATA\r\n"
+         * We expect up to three response lines:
+         *   "DATA\r\n"                 (optional)
          *   "OK <server-id>\r\n"
          *   "AGREE_UNIX_FD\r\n"        (optional)
          */
 
-        d = memmem_safe(b->rbuffer, b->rbuffer_size, "\r\n", 2);
-        if (!d)
-                return 0;
-
-        e = memmem_safe(d + 2, b->rbuffer_size - (d - (char*) b->rbuffer) - 2, "\r\n", 2);
-        if (!e)
-                return 0;
-
-        if (b->accept_fd) {
-                f = memmem_safe(e + 2, b->rbuffer_size - (e - (char*) b->rbuffer) - 2, "\r\n", 2);
-                if (!f)
-                        return 0;
+        n = 0;
+        lines[n] = b->rbuffer;
+        l = memmem_safe(b->rbuffer, b->rbuffer_size, "\r\n", 2);
+        if (l) {
+                lines[++n] = l + 2;
+                l = memmem_safe(l + 2, b->rbuffer_size - (l - (char*) b->rbuffer) - 2, "\r\n", 2);
+                if (l) {
+                        lines[++n] = l + 2;
+                        l = memmem_safe(l + 2, b->rbuffer_size - (l - (char*) b->rbuffer) - 2, "\r\n", 2);
+                        if (l)
+                                lines[++n] = l + 2;
+                }
+        }
 
-                start = f + 2;
-        } else {
-                f = NULL;
-                start = e + 2;
+        /*
+         * In case of ANONYMOUS, we expect OK. In case of EXTERNAL, we expect
+         * DATA+OK. If FD negotiation was requested, we additionally expect
+         * an AGREE_UNIX_FD response.
+         */
+        if (n < (b->anonymous_auth ? 1U : 2U) + !!b->accept_fd)
```

Comment:
If I'm remembering RFC 4422 correctly, `b->anonymous_auth` here is acting as a proxy for: did we send a non-empty initial response in the `AUTH` command? But because systemd currently only implements `EXTERNAL` (*without* an initial response, because systemd wants to send an empty response to the empty challenge, so that it'll work across container boundaries) and `ANONYMOUS` (*with* an initial response, after this MR), that's equivalent to asking whether we're using `ANONYMOUS`.

It would probably make sense to mention that in the comment.
("Did we send"? "Do we plan to send"? Pipelining isn't good at preserving causality.)

Reply:
> It would probably make sense to mention that in the comment.

Mh. I tried to explain that in the comment you quoted. The text matches one-to-one (in-order) the code. It says if ANONYMOUS was used (`b->anonymous_auth == true`), we expect 1 answer. In case of EXTERNAL (`b->anonymous_auth == false`) we expect 2. In both cases, if FD-negotiation was used, an additional answer is expected.

Are you suggesting that `(b->anonymous_auth == false) => EXTERNAL` is not clear? I that case, I can repalce `in case of EXTERNAL` with `in all other cases` maybe?

Reply:
> Are you suggesting that `(b->anonymous_auth == false) => EXTERNAL` is not clear? I that case, I can repalce `in case of EXTERNAL` with `in all other cases` maybe?

No, I'm suggesting that it's not clear that `b->anonymous_auth` is equivalent to "we are using the initial-response optimization".

When you say "inline" in the commit message, what you're really saying is that you're sending a non-empty initial response. See RFC 4422 §3a: `AUTH` is D-Bus' implementation of the "message to initiate the authentication exchange", the first argument is the "field for carrying the name of the mechanism", and the second argument is the "optional field for carrying an initial response".

So perhaps something like this?

```
          * If we sent a non-empty initial response, then we just expect an OK reply.
          * We currently do this if and only if ANONYMOUS.
          * If we did not send an initial response, then we expect the DATA challenge,
          * reply with our own DATA, and expect an OK reply. We do this for EXTERNAL.
          * If FD negotiation was requested, we additionally expect
          * an AGREE_UNIX_FD response. 
```

Reply:
Fair enough, updated!